#
#	modified:   hardware/libhardware_legacy/wifi/Android.mk
#	modified:   hardware/libhardware_legacy/wifi/wifi.c
#
diff --git a/hardware/libhardware_legacy/wifi/Android.mk b/hardware/libhardware_legacy/wifi/Android.mk
index b4a6a7c..10e3cd2 100644
--- a/hardware/libhardware_legacy/wifi/Android.mk
+++ b/hardware/libhardware_legacy/wifi/Android.mk
@@ -15,6 +15,9 @@ endif
 ifdef WIFI_FIRMWARE_LOADER
 LOCAL_CFLAGS += -DWIFI_FIRMWARE_LOADER=\"$(WIFI_FIRMWARE_LOADER)\"
 endif
+ifdef WIFI_DRIVER_LOADER_DELAY
+LOCAL_CFLAGS += -DWIFI_DRIVER_LOADER_DELAY=$(WIFI_DRIVER_LOADER_DELAY)
+endif
 ifdef WIFI_DRIVER_FW_PATH_STA
 LOCAL_CFLAGS += -DWIFI_DRIVER_FW_PATH_STA=\"$(WIFI_DRIVER_FW_PATH_STA)\"
 endif
@@ -29,5 +32,10 @@ LOCAL_CFLAGS += -DWIFI_DRIVER_FW_PATH_PARAM=\"$(WIFI_DRIVER_FW_PATH_PARAM)\"
 endif
 
 LOCAL_SRC_FILES += wifi/wifi.c
+ifeq ($(WIFI_MAC_ADDR_CSPSA),true)
+LOCAL_CFLAGS += -DCSPSA
+LOCAL_C_INCLUDES += device/snda/u8500/include
+LOCAL_SHARED_LIBRARIES += libcspsa
+endif
 
 LOCAL_SHARED_LIBRARIES += libnetutils
diff --git a/hardware/libhardware_legacy/wifi/wifi.c b/hardware/libhardware_legacy/wifi/wifi.c
index 5a42e2d..600b02c 100644
--- a/hardware/libhardware_legacy/wifi/wifi.c
+++ b/hardware/libhardware_legacy/wifi/wifi.c
@@ -87,8 +87,6 @@ static char primary_iface[PROPERTY_VALUE_MAX];
 #define WIFI_DRIVER_FW_PATH_PARAM	"/sys/module/wlan/parameters/fwpath"
 #endif
 
-#define WIFI_DRIVER_LOADER_DELAY	1000000
-
 static const char IFACE_DIR[]           = "/data/system/wpa_supplicant";
 #ifdef WIFI_DRIVER_MODULE_PATH
 static const char DRIVER_MODULE_NAME[]  = WIFI_DRIVER_MODULE_NAME;
@@ -107,6 +105,88 @@ static const char SUPP_CONFIG_FILE[]    = "/data/misc/wifi/wpa_supplicant.conf";
 static const char P2P_CONFIG_FILE[]     = "/data/misc/wifi/p2p_supplicant.conf";
 static const char CONTROL_IFACE_PATH[]  = "/data/misc/wifi/sockets";
 static const char MODULE_FILE[]         = "/proc/modules";
+#ifdef CSPSA
+#include "cspsa.h"
+
+static int wifi_fetch_mac(char* wifi_driver_param)
+{
+    int found = 0;
+    int ret = -1;
+    CSPSA_Result_t result;
+    CSPSA_Handle_t handle;
+    CSPSA_Size_t size;
+    CSPSA_Key_t key = 0x00010000;
+    CSPSA_Data_t *cspsa_data = NULL;
+
+    result = CSPSA_Open("CSPSA0", &handle);
+    if (result != T_CSPSA_RESULT_OK) {
+        ALOGE("Can't open CSPSA area (result 0x%X) ", result);
+        goto cspsa_finished;
+    }
+
+    result = CSPSA_GetSizeOfValue(handle, key, &size);
+    if (result != T_CSPSA_RESULT_OK) {
+        ALOGE("Can't get size of key (h %p key 0x%x result 0x%X).",
+          handle, key, result);
+        goto cspsa_finished;
+    }
+
+    if (size != 6) {
+        ALOGE("Read wrong amount of bytes: %d", size);
+        goto cspsa_finished;
+    }
+
+    cspsa_data = (CSPSA_Data_t *) malloc(size);
+    if (!cspsa_data) {
+        ALOGE("Can't malloc %d bytes.", size);
+        goto cspsa_finished;
+    }
+
+    result = CSPSA_ReadValue(handle, key, size, cspsa_data);
+    if (result != T_CSPSA_RESULT_OK) {
+        ALOGE("Can't read from CSPSA (h %p  key 0x%x size %d res 0x%X).",
+            handle, key, size, result);
+        goto cspsa_finished;
+    }
+
+    /* Don't use CSPSA MAC address if it ends with :00:00:00.*/
+    if ((0 == cspsa_data[3]) && (0 == cspsa_data[4]) && (0 == cspsa_data[5])) {
+        ALOGI("Default MAC address read from CSPSA. Using random.");
+        goto cspsa_finished;
+    }
+
+    found = 1;
+    ret = 0; //success
+    memcpy(wifi_driver_param, cspsa_data, 6);
+    ALOGI("MAC address from CSPSA [0x%02x:0x%02x:0x%02x:0x%02x:0x%02x:0x%02x]",
+         cspsa_data[0], cspsa_data[1], cspsa_data[2],
+         cspsa_data[3], cspsa_data[4], cspsa_data[5]);
+
+cspsa_finished:
+    if (cspsa_data)
+        free(cspsa_data);
+    result = CSPSA_Close(&handle);
+    if (result != T_CSPSA_RESULT_OK)
+        ALOGE("Can't close CSPSA area (result %x) ", result);
+
+    if (found == 0) {
+      memcpy(wifi_driver_param, "macaddr",
+              sizeof(0x00));
+    }
+    return ret;
+}
+
+static void wifi_format_mac_param(char* mac)
+{
+    char mac_address[100];
+    sprintf(mac_address, "0x%02x,0x%02x,0x%02x,0x%02x,0x%02x,0x%02x",
+        mac[0], mac[1], mac[2], mac[3], mac[4], mac[5]);
+    mac[0] = ' '; //one space to separate insmod parameters
+    strcpy(&mac[1], "macaddr");
+    strcat(mac, "=");
+    strcat(mac, mac_address);
+}
+#endif
 
 static const char SUPP_ENTROPY_FILE[]   = WIFI_ENTROPY_FILE;
 static unsigned char dummy_key[21] = { 0x02, 0x11, 0xbe, 0x33, 0x43, 0x35,
@@ -139,8 +219,21 @@ static int insmod(const char *filename, const char *args)
     if (!module)
         return -1;
 
+#ifndef CSPSA
     ret = init_module(module, size, args);
+#else
+    char driver_module_arg[255];
+    char mac_address_param[255];
+    strcpy(driver_module_arg, args);
+
+    if (0 == wifi_fetch_mac(mac_address_param)) {
+        wifi_format_mac_param(mac_address_param);
+    }
+    strcat(driver_module_arg, mac_address_param);
+    ALOGI("***wlan.ko (%s)", driver_module_arg);
 
+    ret = init_module(module, size, driver_module_arg);
+#endif
     free(module);
 
     return ret;
@@ -238,7 +331,9 @@ int wifi_load_driver()
         return -1;
 
     if (strcmp(FIRMWARE_LOADER,"") == 0) {
-        /* usleep(WIFI_DRIVER_LOADER_DELAY); */
+#ifdef WIFI_DRIVER_LOADER_DELAY
+        usleep(WIFI_DRIVER_LOADER_DELAY);
+#endif
         property_set(DRIVER_PROP_NAME, "ok");
     }
     else {
@@ -277,6 +372,7 @@ int wifi_unload_driver()
             usleep(500000);
         }
         usleep(500000); /* allow card removal */
+
         if (count) {
             return 0;
         }
@@ -361,7 +457,6 @@ int update_ctrl_interface(const char *config_file) {
         free(pbuf);
         return 0;
     }
-
     if (!strcmp(config_file, SUPP_CONFIG_FILE)) {
         property_get("wifi.interface", ifc, WIFI_TEST_INTERFACE);
     } else {
